vpr/src/base/vpr_context.h

```diff
@@ -157,6 +169,21 @@ struct DeviceContext : public Context {
     std::vector<std::unique_ptr<ClockNetwork>> clock_networks;
     std::vector<std::unique_ptr<ClockConnection>> clock_connections;
 
+    /** Attributes for each rr_node.
+     * key:     rr_node index
+     * value:   map of <attribute_name, attribute_value>
+     */
+    std::unordered_map<int, t_metadata_dict> rr_node_metadata;
+    /* Attributes for each rr_edge                                             *
+     * key:     <source rr_node_index, sink rr_node_index, iswitch>            *
+     * iswitch: Index of the switch type used to go from this rr_node to       *
+     *          the next one in the routing.  OPEN if there is no next node    *
+     *          (i.e. this node is the last one (a SINK) in a branch of the    *
```

Comment:
I think we should change the key definition here unless there is some reason you need the key definition as is.
It would be simpler and more consistent with our indexing of rr_nodes if it was keyed by 
source_rr_node_index, edge_leaving_source_node_index
That would lead to 
std::unordered_map<std::tuple<int, short>, ...>
The sink node id isn't needed, as you can look it up given the source_rr_node_index and the edge_leaving_source_node_index.  Same for the switch_type.  
This alternative representation is simpler, and keys off the fundamentally unique indices of the rr-graph, so it's better.
It is certainly a strange case, but for example it is electrically legal to have two switches of the same type between two rr_nodes but the current keying could not uniquely identify that case. Better to use a representation that we can't break.

Reply:
There are several reasons the key is the way that it is:
1. Route traceback does not model an iedge, it models a sink_inode and switch. https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/master/vpr/src/base/vpr_types.h#L1201  The traceback is what is stored in the output routing, not the iedge index (which only exists at runtime)
1a. There is no current method to move from the route traceback to an iedge.
2. rr graphs doesn't allow for duplicate edges that share src, sink and switch values https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/master/vpr/src/route/check_rr_graph.cpp#L120
3. iedge values do not exist outside of the runtime, so their stability is in question, and I don't think it makes sense to key off of an unstable value.

For these reasons I believe using sink_inode and switch id is more correct than using the edge in

Reply:
For reference read_route: https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/f166d7c2db35c24bd4f9cd34c100a087dedc75ec/vpr/src/base/read_route.cpp#L320

Reply:
OK, 1 & 2 are good points.  If 3 is in your way, that check is not electrical: it's just a "this looks weird, someone should look at it" error.  Nothing will go wrong in the rest of the program if it is relaxed to an output message (warning, or info).  If you relax that error it would mean you need to figure out what the metadata attaches to in the case that the current 3-tuple key doesn't identify a unique edge (again and odd case, but worth documenting if you decide you need it).